go/enclave/rollupchain/rollup_chain.go

```diff
@@ -878,6 +765,135 @@ func (rc *RollupChain) getRollup(height gethrpc.BlockNumber) (*obscurocore.Rollu
 	return rollup, nil
 }
 
+// Retrieves and encrypts the logs for the block.
+func (rc *RollupChain) getEncryptedLogs(block types.Block, blockState *obscurocore.BlockState) map[gethrpc.ID][]byte {
+	logs := []*types.Log{}
+	fetchedLogs, err := rc.storage.FetchLogs(block.Hash())
+	if err == nil {
+		logs = fetchedLogs
+	} else {
+		rc.logger.Error("Could not retrieve logs for stored block state; returning no logs. Cause: %w", err)
+	}
+	encryptedLogs, err := rc.subscriptionManager.GetSubscribedLogsEncrypted(logs, blockState.HeadRollup)
+	if err != nil {
+		rc.logger.Crit("Could not get subscribed logs in encrypted form. ", log.ErrKey, err)
+	}
+	return encryptedLogs
+}
+
+// Creates a rollup, signs it, checks it, and stores it.
+func (rc *RollupChain) newRollup(block types.Block, blockState *obscurocore.BlockState) (*obscurocore.Rollup, error) {
+	rollup := rc.produceRollup(&block, blockState)
+	rc.signRollup(rollup)
+	// Sanity check the produced rollup
+	_, _, err := rc.checkRollup(rollup)
+	if err != nil {
+		return nil, err
+	}
+
+	// todo - should store proposal rollups in a different storage as they are ephemeral (round based)
+	err = rc.storage.StoreRollup(rollup)
+	if err != nil {
+		return nil, err
+	}
+	return rollup, nil
+}
+
+// Creates a rollup.
+func (rc *RollupChain) produceRollup(b *types.Block, bs *obscurocore.BlockState) *obscurocore.Rollup {
+	headRollup, err := rc.storage.FetchRollup(bs.HeadRollup)
+	if err != nil {
+		rc.logger.Crit("Could not retrieve head rollup", log.ErrKey, err)
+	}
+
+	// These variables will be used to create the new rollup
+	var newRollupTxs []*common.L2Tx
+	var newRollupState *state.StateDB
+
+	/*
```

Comment:
I think we can get rid of this speculative execution stuff now. It mostly makes sense in the context of competing to produce rollups and it's always there in the git history if we want to look back.